daemon/container.go

```diff
@@ -261,14 +261,44 @@ func populateCommand(c *Container, env []string) error {
 	pid.HostPid = c.hostConfig.PidMode.IsHost()
 
 	// Build lists of devices allowed and created within the container.
-	userSpecifiedDevices := make([]*configs.Device, len(c.hostConfig.Devices))
-	for i, deviceMapping := range c.hostConfig.Devices {
+	var userSpecifiedDevices []*configs.Device
+	for _, deviceMapping := range c.hostConfig.Devices {
 		device, err := devices.DeviceFromPath(deviceMapping.PathOnHost, deviceMapping.CgroupPermissions)
+		// if the device is not a device node
+		// let's try to see if it's a directory holding many devices
+		if err == devices.ErrNotADevice {
+			// check if it is a directory
+			if src, e := os.Stat(deviceMapping.PathOnHost); e == nil && src.IsDir() {
+				// mount the internal devices recursively
+				var foundChildDevice bool
+				filepath.Walk(deviceMapping.PathOnHost, func(dpath string, f os.FileInfo, e error) error {
+					childDevice, e := devices.DeviceFromPath(dpath, deviceMapping.CgroupPermissions)
+					if e != nil {
+						// ignore the device
+						return nil
+					}
+
+					// add the device to userSpecified devices
+					foundChildDevice = true
+					childDevice.Path = strings.Replace(dpath, deviceMapping.PathOnHost, deviceMapping.PathInContainer, 1)
+					userSpecifiedDevices = append(userSpecifiedDevices, childDevice)
+
+					return nil
+				})
+
+				if foundChildDevice {
```

Comment:
Ya i think i am going to redo to use new libcontiainer api stuffs, just havent found time